NODE-446 /node/status returns state height that lags behind blockchai…

petermz opened the pull request:
JIRA: https://wavesplatform.atlassian.net/browse/NODE-446

Instead of reading persisted state height from StateWriter, I'm now reading complete (persisted + in-memory) state height from SnapshotStateReader

ismagin commented on src/main/scala/com/wavesplatform/state2/StateStorage.scala:
```diff
@@ -17,9 +17,14 @@ class StateStorage private(file: Option[File]) extends VariablesStorage(createMV
 
   override protected val Version = 2
 
+  @volatile private[state2] var lastUpdated: Long = 0L
```
Please clean this up, make a public getter and private setter, package-private vars is a mess

ismagin commented on src/main/scala/com/wavesplatform/state2/StateWriter.scala:
```diff
@@ -5,22 +5,18 @@ import java.util.concurrent.locks.ReentrantReadWriteLock
   def clear(): Unit
 
-  def status: Observable[Status]
+  def lastUpdated: Long
```
is `Status` not used anymore?

ismagin commented on src/main/scala/com/wavesplatform/state2/StateStorage.scala:
```diff
@@ -17,9 +17,14 @@ class StateStorage private(file: Option[File]) extends VariablesStorage(createMV
-  def setHeight(i: Int): Unit = putInt(heightKey, i)
+  def setHeight(i: Int): Unit = {
+    putInt(heightKey, i)
+    lastUpdated = System.currentTimeMillis
```
Why not use `Time`? It is not critical here, but it's best if we use single source of time everywhere

ismagin commented on src/main/scala/com/wavesplatform/http/NodeApiRoute.scala:
```diff
@@ -49,7 +49,7 @@ case class NodeApiRoute(settings: RestAPISettings, blockchainUpdater: Blockchain
   def status: Route = (get & path("status")) {
     val (bcHeight, bcTime) = lastHeight().map { case LastBlockInfo(_, h, _, _, t) => (h, t) }.getOrElse((0, 0L))
-    val (stHeight, stTime) = lastState().map { case StateWriter.Status(h, t) => (h, t) }.getOrElse((0, 0L))
+    val (stHeight, stTime) = (stateReader().height, stateWriter.lastUpdated)
```
There's th `trait BlockchainDebugInfo`, please use it this stuff, BlockchainUpdater and StateWriter shouldn't be here at all

petermz commented:
In this commit:
- Moved stuff into BlockchainDebugInfo to avoid exposing StateReader/Writer
- Added history height and timestamps to StateDebugInfo (should I rename this class now?)
- Time is used everywhere instead of System.currentTimeMillis

petermz commented:
No, wait, that doesn't work because of locking =(

ismagin commented:
Looks like all you need to expose from not-yet-rebuilt state is height. Maybe you can just expose coeval instead of  accepting callback?

ismagin commented:
NODE-467 Stop exposing internal entities to NodeApiRoute #780 suites better
